data_iterator.py

```diff
@@ -51,22 +56,35 @@ def create_padded_batch(state, x):
             X[triple_length:, idx] = state['eos_sym']
 
         # Initialize Xmask column with ones in all positions that
-        # were just set in X
+        # were just set in X. 
+        # Note: if we need mask to depend on tokens inside X, then we need to 
+        # create a corresponding mask for X_reversed and send it further in the model
         Xmask[:triple_length, idx] = 1.
 
-        # Find start of last utterance
+        # Reverse all utterances
         eos_indices = numpy.where(X[:, idx] == state['eos_sym'])[0]
+        X_reversed[:triple_length, idx] = x[0][idx][:triple_length]
+        prev_eos_index = -1
+        for eos_index in eos_indices:
+            X_reversed[(prev_eos_index+2):eos_index, idx] = (X_reversed[(prev_eos_index+2):eos_index, idx])[::-1]
+            prev_eos_index = eos_index
+            if prev_eos_index > triple_length:
+                break
```

Comment:
Did you notice slowdowns? Preferably, it may be better to create reversed_data_iterator.py and put the reverse code there. Otw, even if we want to run only forward one, this gets executed?